In [1]:
import pandas as pd
import numpy as np
import requests
import json
import tweepy

In [2]:
consumer_key = 'VlFaFpY1w2oaWTN3lItGykJS4'
consumer_secret = 'It5cUQJ6CWBF13n6AUpHx0eHslR5r0NXIQLkebDKGMrfkuaSWv'
access_token = '2685226321-7KE4KEqMLCZeht5tcRxNfCJ8Iduo9K4feEsX5DR'
access_secret = 'QkPPBiJaLBtkh4zokD1qVtBiV5DBA2a21wZM6OkpWIWGW'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

## 1- Data Gathering

In [3]:
# create the first dataframe to store twitter-archive-enhanced.csv file
df_twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [4]:
# downloading image_predictions.tsv file programmatically
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

#writing the content of response to a file
with open(url.split('/')[-1], mode='wb') as file:
    file.write(response.content)
    
# create the second dataframe to store the image_predications.tsv file
df_dog_breeds = pd.read_csv('image-predictions.tsv', sep='\t')

In [5]:
# store tweets data in tweet_json.txt file

fails_dict= {} # tweets we couldn't find
with open ('tweet_json.txt', 'w') as file:
    for tweet_id in df_twitter_archive.tweet_id:
        try:
            tweet = api.get_status(tweet_id, tweet_mode = 'extended')
            json.dump(tweet._json, file)
            file.write('\n')
        except tweepy.TweepError as e:
            fails_dict[tweet_id] = e
            pass

In [6]:
# read the "tweet_json.txt" line be line and
# each line is a dictionary, so can access the data we need
# in each line and store in list. After we finish, we will
# convert the list of dictionary into a datafram

tweets_list = []
with open('tweet_json.txt', 'r') as json_file:
    for line in json_file:
        json_line = json.loads(line)
        tweets_list.append({
            'tweet_id': json_line['id'],
            'created_at': json_line['created_at'],
            'retweet_count': json_line['retweet_count'],
            'favorite_count': json_line['favorite_count']
        })

In [7]:
# create the third dataframe df_tweets
df_tweets = pd.DataFrame(tweets_list, columns=['tweet_id', 'retweet_count', 'favorite_count', 'created_at'])

In [8]:
df_twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [9]:
df_dog_breeds

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [10]:
df_tweets

tweet_id  retweet_count  favorite_count  \
0    892420643555336193           7696           36134   
1    892177421306343426           5691           31164   
2    891815181378084864           3775           23477   
3    891689557279858688           7860           39421   
4    891327558926688256           8468           37648   
5    891087950875897856           2837           18967   
6    890971913173991426           1846           11034   
7    890729181411237888          17170           60859   
8    890609185150312448           3907           26094   
9    890240255349198849           6679           29841   
10   890006608113172480           6664           28719   
11   889880896479866881           4541           26130   
12   889665388333682689           9105           44974   
13   889638837579907072           4080           25319   
14   889531135344209921           2054           14176   
15   889278841981685760           4854           23596   
16   888917238123831296           4097           27229   
17   888804989199671297           3861           23953   
18   888554962724278272           3171           18486   
19   888078434458587136           3158           20400   
20   887705289381826560           4905           28308   
21   887517139158093824          10723           43391   
22   887473957103951883          16373           64330   
23   887343217045368832           9536           31504   
24   887101392804085760           5431           28655   
25   886983233522544640           6986           32672   
26   886736880519319552           2930           11224   
27   886680336477933568           4076           21041   
28   886366144734445568           2885           19778   
29   886267009285017600              4             110   
..                  ...            ...             ...   
847  761745352076779520            869            4293   
848  761672994376806400          29300           49769   
849  761599872357261312           1194            4144   
850  761371037149827077          18179               0   
851  761334018830917632           1463            5217   
852  761292947749015552           1091            4506   
853  761227390836215808           1575            5345   
854  761004547850530816           3498           11383   
855  760893934457552897            985            3827   
856  760656994973933572           1925            6606   
857  760641137271070720           1309            4997   
858  760539183865880579           3650            7593   
859  760521673607086080           1393            4234   
860  760290219849637889          11565           26921   
861  760252756032651264            871            3957   
862  760190180481531904           1818            5720   
863  760153949710192640             31               0   
864  759943073749200896           2090            5975   
865  759923798737051648           5686           14731   
866  759846353224826880           1986            6759   
867  759793422261743616           1926            5984   
868  759557299618865152           1194            4719   
869  759447681597108224           2470            8522   
870  759446261539934208            492            1679   
871  759197388317847553           1956            6095   
872  759159934323924993           1162               0   
873  759099523532779520           4192           14577   
874  759047813560868866           2032            6515   
875  758854675097526272            902            3559   
876  758828659922702336           3854           11250   

                         created_at  
0    Tue Aug 01 16:23:56 +0000 2017  
1    Tue Aug 01 00:17:27 +0000 2017  
2    Mon Jul 31 00:18:03 +0000 2017  
3    Sun Jul 30 15:58:51 +0000 2017  
4    Sat Jul 29 16:00:24 +0000 2017  
5    Sat Jul 29 00:08:17 +0000 2017  
6    Fri Jul 28 16:27:12 +0000 2017  
7    Fri Jul 28 00:22:40 +0000 2017  
8    Thu Jul 27 16:25:51 +0000 2017  
9    Wed Jul 26 15:59:51 +0000 2017  


## 2- Assess Data

In [11]:
df_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

- timestamp shouldn't of type oject
- missing values in mutiple columns (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted, status_timestamp)
- dog stages are in four columns

In [12]:
df_dog_breeds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


- tweet_id should of type object not int

In [13]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 877 entries, 0 to 876
Data columns (total 4 columns):
tweet_id          877 non-null int64
retweet_count     877 non-null int64
favorite_count    877 non-null int64
created_at        877 non-null object
dtypes: int64(3), object(1)
memory usage: 27.5+ KB


- tweet_id column should be string not int
- created_at column should be date not string.

In [14]:
df_twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [15]:
df_twitter_archive[['tweet_id','text', 'rating_numerator','rating_denominator', 'name']]


tweet_id                                               text  \
0     892420643555336193  This is Phineas. He's a mystical boy. Only eve...   
1     892177421306343426  This is Tilly. She's just checking pup on you....   
2     891815181378084864  This is Archie. He is a rare Norwegian Pouncin...   
3     891689557279858688  This is Darla. She commenced a snooze mid meal...   
4     891327558926688256  This is Franklin. He would like you to stop ca...   
5     891087950875897856  Here we have a majestic great white breaching ...   
6     890971913173991426  Meet Jax. He enjoys ice cream so much he gets ...   
7     890729181411237888  When you watch your owner call another dog a g...   
8     890609185150312448  This is Zoey. She doesn't want to be one of th...   
9     890240255349198849  This is Cassie. She is a college pup. Studying...   
10    890006608113172480  This is Koda. He is a South Australian decksha...   
11    889880896479866881  This is Bruno. He is a service shark. Only get...   
12    889665388333682689  Here's a puppo that seems to be on the fence a...   
13    889638837579907072  This is Ted. He does his best. Sometimes that'...   
14    889531135344209921  This is Stuart. He's sporting his favorite fan...   
15    889278841981685760  This is Oliver. You're witnessing one of his m...   
16    888917238123831296  This is Jim. He found a fren. Taught him how t...   
17    888804989199671297  This is Zeke. He has a new stick. Very proud o...   
18    888554962724278272  This is Ralphus. He's powering up. Attempting ...   
19    888202515573088257  RT @dog_rates: This is Canela. She attempted s...   
20    888078434458587136  This is Gerald. He was just told he didn't get...   
21    887705289381826560  This is Jeffrey. He has a monopoly on the pool...   
22    887517139158093824  I've yet to rate a Venezuelan Hover Wiener. Th...   
23    887473957103951883  This is Canela. She attempted some fancy porch...   
24    887343217045368832  You may not have known you needed to see this ...   
25    887101392804085760  This... is a Jubilant Antarctic House Bear. We...   
26    886983233522544640  This is Maya. She's very shy. Rarely leaves he...   
27    886736880519319552  This is Mingus. He's a wonderful father to his...   
28    886680336477933568  This is Derek. He's late for a dog meeting. 13...   
29    886366144734445568  This is Roscoe. Another pupper fallen victim t...   
...                  ...                                                ...   
2326  666411507551481857  This is quite the dog. Gets really excited whe...   
2327  666407126856765440  This is a southern Vesuvius bumblegruff. Can d...   
2328  666396247373291520  Oh goodness. A super rare northeast Qdoba kang...   
2329  666373753744588802  Those are sunglasses and a jean jacket. 11/10 ...   
2330  666362758909284353  Unique dog here. Very small. Lives in containe...   
2331  666353288456101888  Here we have a mixed Asiago from the Galápagos...   
2332  666345417576210432  Look at this jokester thinking seat belt laws ...   
2333  666337882303524864  This is an extremely rare horned Parthenon. No...   
2334  666293911632134144  This is a funny dog. Weird toes. Won't come do...   
2335  666287406224695296  This is an Albanian 3 1/2 legged  Episcopalian...   
2336  666273097616637952     Can take selfies 11/10 https://t.co/ws2AMaNwPW   
2337  666268910803644416  Very concerned about fellow dog trapped in com...   
2338  666104133288665088  Not familiar with this breed. No tail (weird)....   
2339  666102155909144576  Oh my. Here you are seeing an Adobe Setter giv...   
2340  666099513787052032  Can stand on stump for what seems like a while...   
2341  666094000022159362  This appears to be a Mongolian Presbyterian mi...   
2342  666082916733198337  Here we have a well-established sunblockerspan...   
2343  666073100786774016  Let's hope this flight isn't Malaysian (lol). ...   
2344  666071193221509120  Here we have a northern speckled Rhodode

- at row 2335 rating_numerator and rating_denominator are extrated wrong. they should be 9 and 10 instead of 1 and 2

In [16]:
df_twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

- some values in the rating_denominator are greater than 10

In [17]:
df_twitter_archive.name.value_counts()

None            745
a                55
Charlie          12
Cooper           11
Lucy             11
Oliver           11
Lola             10
Penny            10
Tucker           10
Bo                9
Winston           9
the               8
Sadie             8
Toby              7
Bailey            7
an                7
Daisy             7
Buddy             7
Jack              6
Koda              6
Oscar             6
Stanley           6
Bella             6
Jax               6
Dave              6
Milo              6
Scout             6
Leo               6
Rusty             6
Oakley            5
               ... 
Benny             1
Ed                1
Darrel            1
Zuzu              1
Alexanderson      1
Maxwell           1
Jiminus           1
Charl             1
Marq              1
Jeffrie           1
Dallas            1
Callie            1
Ralphus           1
Margo             1
Kial              1
Acro              1
Gòrdón            1
Stuart            1
Marty             1


 - there are values in the name column extracted wrong. As you can see above, there are 745 values = "None", 55 = the "a" letter. 

In [18]:
df_twitter_archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

- outliers in the rating_numerator column

In [19]:
df_twitter_archive[['tweet_id','text', 'rating_numerator','rating_denominator']][df_twitter_archive.rating_numerator > 14]

tweet_id                                               text  \
55    881633300179243008  @roushfenway These are good dogs but 17/10 is ...   
188   855862651834028034  @dhmontgomery We also gave snoop dogg a 420/10...   
189   855860136149123072  @s8n You tried very hard to portray this good ...   
285   838916489579200512  RT @KibaDva: I collected all the good dogs!! 1...   
290   838150277551247360                                 @markhoppus 182/10   
291   838085839343206401  @bragg6of8 @Andy_Pace_ we are still looking fo...   
313   835246439529840640  @jonnysun @Lin_Manuel ok jomny I know you're e...   
340   832215909146226688  RT @dog_rates: This is Logan, the Chow who liv...   
433   820690176645140481  The floofs have been released I repeat the flo...   
516   810984652412424192  Meet Sam. She smiles 24/7 &amp; secretly aspir...   
695   786709082849828864  This is Logan, the Chow who lived. He solemnly...   
763   778027034220126208  This is Sophie. She's a Jubilant Bush Pupper. ...   
902   758467244762497024  Why does this never happen at my front door......   
979   749981277374128128  This is Atticus. He's quite simply America af....   
1120  731156023742988288  Say hello to this unbelievably well behaved sq...   
1202  716439118184652801  This is Bluebert. He just saw that both #Final...   
1228  713900603437621249  Happy Saturday here's 9 puppers on a bench. 99...   
1254  710658690886586372  Here's a brigade of puppers. All look very pre...   
1274  709198395643068416  From left to right:\nCletus, Jerome, Alejandro...   
1351  704054845121142784  Here is a whole flock of puppers.  60/50 I'll ...   
1433  697463031882764288  Happy Wednesday here's a bucket of pups. 44/40...   
1634  684225744407494656  Two sneaky puppers were not initially seen, mo...   
1635  684222868335505415  Someone help the girl is being mugged. Several...   
1663  682808988178739200  I'm aware that I could've said 20/16, but here...   
1712  680494726643068929  Here we have uncovered an entire battalion of ...   
1779  677716515794329600  IT'S PUPPERGEDDON. Total of 144/120 ...I think...   
1843  675853064436391936  Here we have an entire platoon of puppers. Tot...   
2074  670842764863651840  After so many requests... here you go.\n\nGood...   

      rating_numerator  rating_denominator  
55                  17                  10  
188                420                  10  
189                666                  10  
285                 15                  10  
290                182                  10  
291                 15                  10  
313                960                   0  
340                 75                  10  
433                 84                  70  
516                 24                   7  
695                 75                  10  
763                 27                  10  
902                165                 150  
979               1776                  10  
1120               204                 170  
1202                50                  50  
1228                99                  90  
1254                80                  80  
1274                45                  50  
1351                60                  50  
1433                44                  40  
1634               143                 130  
1635               121                 110  
1663                20                  16  
1712                26                  10  
1779               144                 120  
1843                88                  80  
2074               420                  10

- decimal ratings aren't extracted properly (tweet_id = 832215909146226688)

### Data Quality Issues
- tweet_id in the three dataframe is int.It shold be string
- timestamp in the df_arhcive is string
- missing values in mutiple columns (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted, status_timestamp)
- some values in the rating_denominator are greater than 10
- retweets in the df_archive table
- there are values in the name column extracted wrong from the tweet's text
- outliers in the rating_numerator column
- At row 2335 in the df_archive table rating_numerator and rating_denominator are extrated wrong. they should be 9 and 10 instead of 1 and 2
- decimal ratings aren't extracted properly (tweet_id = 832215909146226688)
- ratings should be float ranging since we have decimal values



### Data Tidiness Issues
- dog stages are in four columns
- tweets info separated in two tables (df_archive_tweets and df_tweets)
- dog breed should be a column in the combined table

## 3- Data Cleaning

In [20]:
# create a copy df_twitter_archiver. This copy will be the cleaned version 
df_twitter_archive_cleaned = df_twitter_archive.copy()

In [21]:
df_tweets_clean = df_tweets.copy()
df_dog_breeds_clean = df_dog_breeds.copy()

### Define
- remove the retweets data from the df_twitter_archive table

### Code

In [22]:
# fitering out retweets
df_twitter_archive_cleaned = df_twitter_archive_cleaned[df_twitter_archive.retweeted_status_id.isna()]

### test

In [23]:
df_twitter_archive[df_twitter_archive.name == 'Canela']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19  888202515573088257                    NaN                  NaN   
23  887473957103951883                    NaN                  NaN   

                    timestamp  \
19  2017-07-21 01:02:36 +0000   
23  2017-07-19 00:47:34 +0000   

                                               source  \
19  <a href="http://twitter.com/download/iphone" r...   
23  <a href="http://twitter.com/download/iphone" r...   

                                                 text  retweeted_status_id  \
19  RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
23  This is Canela. She attempted some fancy porch...                  NaN   

    retweeted_status_user_id retweeted_status_timestamp  \
19              4.196984e+09  2017-07-19 00:47:34 +0000   
23                       NaN                        NaN   

                                        expanded_urls  rating_numerator  \
19  https://twitter.com/dog_rates/status/887473957...                13   
23  https://twitter.com/dog_rates/status/887473957...                13   

    rating_denominator    name doggo floofer pupper puppo  
19                  10  Canela  None    None   None  None  
23                  10  Canela  None    None   None  None

In [24]:
df_twitter_archive_cleaned[df_twitter_archive_cleaned.name == 'Canela']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
23  887473957103951883                    NaN                  NaN   

                    timestamp  \
23  2017-07-19 00:47:34 +0000   

                                               source  \
23  <a href="http://twitter.com/download/iphone" r...   

                                                 text  retweeted_status_id  \
23  This is Canela. She attempted some fancy porch...                  NaN   

    retweeted_status_user_id retweeted_status_timestamp  \
23                       NaN                        NaN   

                                        expanded_urls  rating_numerator  \
23  https://twitter.com/dog_rates/status/887473957...                13   

    rating_denominator    name doggo floofer pupper puppo  
23                  10  Canela  None    None   None  None

#### As you can see above, the original dataframe has the original tweet and the retweet. The cleaned dataframe doesn't have the retweet

In [25]:
# check if there are data in the retweeted_status_id column
df_twitter_archive_cleaned[~df_twitter_archive_cleaned.retweeted_status_id.isna()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [26]:
# check if there are data in the retweeted_status_user_id column
df_twitter_archive_cleaned[~df_twitter_archive_cleaned.retweeted_status_user_id.isna()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [27]:
# check if there are data in the rretweeted_status_timestamp column
df_twitter_archive_cleaned[~df_twitter_archive_cleaned.retweeted_status_timestamp.isna()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

### Missing data

In [28]:
df_twitter_archive_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

### Define
- Remove missing data
- we can drop retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp since they don't have any data
- we can also drop in_reply_to_status_id and in_reply_to_user_id since they a lot of missing data and we won't use them ain our analysis 

### Code

In [29]:
# drop columns with missing values
df_twitter_archive_cleaned.drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp',
                                 'in_reply_to_status_id', 'in_reply_to_user_id'], axis=1, inplace = True)

### Test

In [30]:
df_twitter_archive_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: int64(3), object(9)
memory usage: 220.9+ KB


### Define
- Create one variable/column for the four dog stages

### Code

In [31]:
# reaplce None with empty string in the four columns
df_twitter_archive_cleaned.doggo.replace('None','', inplace=True)
df_twitter_archive_cleaned.floofer.replace('None','', inplace=True)
df_twitter_archive_cleaned.pupper.replace('None','', inplace=True)
df_twitter_archive_cleaned.puppo.replace('None','', inplace=True)
df_twitter_archive_cleaned[['text', 'doggo', 'floofer', 'pupper', 'puppo']]

text  doggo floofer  \
0     This is Phineas. He's a mystical boy. Only eve...                  
1     This is Tilly. She's just checking pup on you....                  
2     This is Archie. He is a rare Norwegian Pouncin...                  
3     This is Darla. She commenced a snooze mid meal...                  
4     This is Franklin. He would like you to stop ca...                  
5     Here we have a majestic great white breaching ...                  
6     Meet Jax. He enjoys ice cream so much he gets ...                  
7     When you watch your owner call another dog a g...                  
8     This is Zoey. She doesn't want to be one of th...                  
9     This is Cassie. She is a college pup. Studying...  doggo           
10    This is Koda. He is a South Australian decksha...                  
11    This is Bruno. He is a service shark. Only get...                  
12    Here's a puppo that seems to be on the fence a...                  
13    This is Ted. He does his best. Sometimes that'...                  
14    This is Stuart. He's sporting his favorite fan...                  
15    This is Oliver. You're witnessing one of his m...                  
16    This is Jim. He found a fren. Taught him how t...                  
17    This is Zeke. He has a new stick. Very proud o...                  
18    This is Ralphus. He's powering up. Attempting ...                  
20    This is Gerald. He was just told he didn't get...                  
21    This is Jeffrey. He has a monopoly on the pool...                  
22    I've yet to rate a Venezuelan Hover Wiener. Th...                  
23    This is Canela. She attempted some fancy porch...                  
24    You may not have known you needed to see this ...                  
25    This... is a Jubilant Antarctic House Bear. We...                  
26    This is Maya. She's very shy. Rarely leaves he...                  
27    This is Mingus. He's a wonderful father to his...                  
28    This is Derek. He's late for a dog meeting. 13...                  
29    This is Roscoe. Another pupper fallen victim t...                  
30    @NonWhiteHat @MayhewMayhem omg hello tanner yo...                  
...                                                 ...    ...     ...   
2326  This is quite the dog. Gets really excited whe...                  
2327  This is a southern Vesuvius bumblegruff. Can d...                  
2328  Oh goodness. A super rare northeast Qdoba kang...                  
2329  Those are sunglasses and a jean jacket. 11/10 ...                  
2330  Unique dog here. Very small. Lives in containe...                  
2331  Here we have a mixed Asiago from the Galápagos...                  
2332  Look at this jokester thinking seat belt laws ...                  
2333  This is an extremely rare horned Parthenon. No...                  
2334  This is a funny dog. Weird toes. Won't come do...                  
2335  This is an Albanian 3 1/2 legged  Episcopalian...                  
2336     Can take selfies 11/10 https://t.co/ws2AMaNwPW                  
2337  Very concerned about fellow dog trapped in com...                  
2338  Not familiar with this breed. No tail (weird)....                  
2339  Oh my. Here you are seeing an Adobe Setter giv...                  
2340  Can stand on stump for what seems like a while...                  
2341  This appears to be a Mongolian Presbyterian mi...                  
2342  Here we have a well-established sunblockerspan...                  
2343  Let's hope this flight isn't Malaysian (lol). ...                  
2344  Here we have a northern speckled Rhododendron....                  
2345  This is the happiest dog you will ever see. Ve...                  
2346  Here is the Rand Paul of retrievers folks! He'...                  
2347  My oh my. This is a rare blond Canadian terrie...                  
2348  Here is a Siberian heavily armored polar bear ...

In [32]:
# create a new column (dog_stage) for the dog stages
df_twitter_archive_cleaned['dog_stage'] = df_twitter_archive_cleaned.doggo + df_twitter_archive_cleaned.floofer + df_twitter_archive_cleaned.pupper + df_twitter_archive_cleaned.puppo

In [33]:
df_twitter_archive_cleaned.dog_stage.value_counts()

                1831
pupper           224
doggo             75
puppo             24
doggopupper       10
floofer            9
doggofloofer       1
doggopuppo         1
Name: dog_stage, dtype: int64

#### As you can see above there are some dogs with muliple stages.  we can sperate those instances with commas

In [34]:
df_twitter_archive_cleaned.loc[df_twitter_archive_cleaned.dog_stage == 'doggopupper', 'dog_stage'] = 'doggo, pupper'
df_twitter_archive_cleaned.loc[df_twitter_archive_cleaned.dog_stage == 'doggofloofer', 'dog_stage'] = 'doggo, floofer'
df_twitter_archive_cleaned.loc[df_twitter_archive_cleaned.dog_stage == 'doggopuppo', 'dog_stage'] = 'doggo, puppo'

In [35]:
df_twitter_archive_cleaned.dog_stage.value_counts()

                  1831
pupper             224
doggo               75
puppo               24
doggo, pupper       10
floofer              9
doggo, floofer       1
doggo, puppo         1
Name: dog_stage, dtype: int64

#### we can now convert all the empty strings to NaN

In [36]:
df_twitter_archive_cleaned.dog_stage.replace('', np.nan, inplace=True)

#### Now we have one column that represents the dog stages. we can delete the other four columns

In [37]:
df_twitter_archive_cleaned.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1, inplace=True)

### Test

In [38]:
df_twitter_archive_cleaned.dog_stage.value_counts()

pupper            224
doggo              75
puppo              24
doggo, pupper      10
floofer             9
doggo, floofer      1
doggo, puppo        1
Name: dog_stage, dtype: int64

In [39]:
df_twitter_archive_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
dog_stage             344 non-null object
dtypes: int64(3), object(6)
memory usage: 169.9+ KB


### Define 
- Merge df_twitter_archive_cleaned and df_tweets_cleaned in a new table (df_all_tweets)
- Before we merge the two tables let change the type of tweet_id in both tables to string/object
- I decided to delete the "created_at" in the df_tweets table since it is a duplicate

### Code

In [40]:
# change the type of tweet_id in both tables to string/object
df_twitter_archive_cleaned.tweet_id = df_twitter_archive_cleaned.tweet_id.astype(str)
df_tweets_clean.tweet_id = df_tweets_clean.tweet_id.astype(str)
df_tweets_clean .drop('created_at', axis=1, inplace=True)

In [41]:
df_all_tweets = df_twitter_archive_cleaned.merge(df_tweets_clean, on='tweet_id', how='left')

### Test

In [42]:
df_all_tweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2174
Data columns (total 11 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
dog_stage             344 non-null object
retweet_count         726 non-null float64
favorite_count        726 non-null float64
dtypes: float64(2), int64(2), object(7)
memory usage: 203.9+ KB


### Define
- change type of tweet_id in df_dog_breeds_clean table to string/object
- create a new column "dog_breed" in df_dog_breeds table.
- merge df_all_tweets and df_dog_breeds_clean

In [43]:
# change type of tweet_id in df_dog_breeds_clean table to string/object
df_dog_breeds_clean.tweet_id = df_dog_breeds_clean.tweet_id.astype(str)

In [44]:
# function below is responsible for populating the new column "dog_breed"
def get_dog_breed(df):
    if (df.p1_dog == True):
        return df.p1
    elif (df.p2_dog == True):
        return df.p2
    elif (df.p3_dog == True):
        return df.p3
    else:
        return np.nan

In [45]:
df_dog_breeds_clean['dog_breed'] = df_dog_breeds_clean.apply(get_dog_breed, axis=1)

In [53]:
df_dog_breeds_clean

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [50]:
df_all_tweets = df_all_tweets.merge(df_dog_breeds_clean[['tweet_id','dog_breed','jpg_url']], on='tweet_id', how='left')

### Test

In [51]:
df_all_tweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2174
Data columns (total 13 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
dog_stage             344 non-null object
retweet_count         726 non-null float64
favorite_count        726 non-null float64
dog_breed             1686 non-null object
jpg_url               1994 non-null object
dtypes: float64(2), int64(2), object(9)
memory usage: 237.9+ KB


In [52]:
df_all_tweets

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
5     891087950875897856  2017-07-29 00:08:17 +0000   
6     890971913173991426  2017-07-28 16:27:12 +0000   
7     890729181411237888  2017-07-28 00:22:40 +0000   
8     890609185150312448  2017-07-27 16:25:51 +0000   
9     890240255349198849  2017-07-26 15:59:51 +0000   
10    890006608113172480  2017-07-26 00:31:25 +0000   
11    889880896479866881  2017-07-25 16:11:53 +0000   
12    889665388333682689  2017-07-25 01:55:32 +0000   
13    889638837579907072  2017-07-25 00:10:02 +0000   
14    889531135344209921  2017-07-24 17:02:04 +0000   
15    889278841981685760  2017-07-24 00:19:32 +0000   
16    888917238123831296  2017-07-23 00:22:39 +0000   
17    888804989199671297  2017-07-22 16:56:37 +0000   
18    888554962724278272  2017-07-22 00:23:06 +0000   
19    888078434458587136  2017-07-20 16:49:33 +0000   
20    887705289381826560  2017-07-19 16:06:48 +0000   
21    887517139158093824  2017-07-19 03:39:09 +0000   
22    887473957103951883  2017-07-19 00:47:34 +0000   
23    887343217045368832  2017-07-18 16:08:03 +0000   
24    887101392804085760  2017-07-18 00:07:08 +0000   
25    886983233522544640  2017-07-17 16:17:36 +0000   
26    886736880519319552  2017-07-16 23:58:41 +0000   
27    886680336477933568  2017-07-16 20:14:00 +0000   
28    886366144734445568  2017-07-15 23:25:31 +0000   
29    886267009285017600  2017-07-15 16:51:35 +0000   
...                  ...                        ...   
2145  666411507551481857  2015-11-17 00:24:19 +0000   
2146  666407126856765440  2015-11-17 00:06:54 +0000   
2147  666396247373291520  2015-11-16 23:23:41 +0000   
2148  666373753744588802  2015-11-16 21:54:18 +0000   
2149  666362758909284353  2015-11-16 21:10:36 +0000   
2150  666353288456101888  2015-11-16 20:32:58 +0000   
2151  666345417576210432  2015-11-16 20:01:42 +0000   
2152  666337882303524864  2015-11-16 19:31:45 +0000   
2153  666293911632134144  2015-11-16 16:37:02 +0000   
2154  666287406224695296  2015-11-16 16:11:11 +0000   
2155  666273097616637952  2015-11-16 15:14:19 +0000   
2156  666268910803644416  2015-11-16 14:57:41 +0000   
2157  666104133288665088  2015-11-16 04:02:55 +0000   
2158  666102155909144576  2015-11-16 03:55:04 +0000   
2159  666099513787052032  2015-11-16 03:44:34 +0000   
2160  666094000022159362  2015-11-16 03:22:39 +0000   
2161  666082916733198337  2015-11-16 02:38:37 +0000   
2162  666073100786774016  2015-11-16 01:59:36 +0000   
2163  666071193221509120  2015-11-16 01:52:02 +0000   
2164  666063827256086533  2015-11-16 01:22:45 +0000   
2165  666058600524156928  2015-11-16 01:01:59 +0000   
2166  666057090499244032  2015-11-16 00:55:59 +0000   
2167  666055525042405380  2015-11-16 00:49:46 +0000   
2168  666051853826850816  2015-11-16 00:35:11 +0000   
2169  666050758794694657  2015-11-16 00:30:50 +0000   
2170  666049248165822465  2015-11-16 00:24:50 +0000   
2171  666044226329800704  2015-11-16 00:04:52 +0000   
2172  666033412701032449  2015-11-15 23:21:54 +0000   
2173  666029285002620928  2015-11-15 23:05:30 +0000   
2174  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="